### PR TITLE
rename fast_dev_run -> unit_test

### LINKED ISSUE #1081: fast_dev_run -> unit_test
Anyone want to make this change?

- rename fast_dev_run -> unit_test
- add checking the test set as well (currently only checks val, train).

### PR DESCRIPTION
**Pull request by Gokkulnath**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements) 
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section? Yes
- [x] Did you make sure to update the docs?   Yes
- [x] Did you write any new necessary tests?  Not Applicable
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CHANGELOG.md)? I can update if required. It is a minor change. 

## What does this PR do?
Fixes #1081

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
hey there, we have added GPU CI test, so could we kindly ask to rebase/merge master which will trigger these tests so we do not need to test it manually... Thx for your understanding :robot:

File: pytorch_lightning/trainer/__init__.py [hunk 1]
```diff
@@ -270,7 +270,7 @@ def on_train_end(self):
 
 .. note:: If ``'val_loss'`` is not found will work as if early stopping is disabled.
 
-fast_dev_run
+unit_test
 ^^^^^^^^^^^^
```
Comment (Borda) [inline pytorch_lightning/trainer/__init__.py]:
```suggestion
^^^^^^^^^
```

**Comment by Borda** [comment]
@Gokkulnath mind rebase and check the comments above? :]

**Comment by Gokkulnath** [comment]
@Borda  I am not sure why the CI testing is failing for ubuntu and Mac Os can you please help me in resolving the issue ?

**Comment by williamFalcon** [comment]
@Gokkulnath the rename is good. But this PR also needs to add a run through a test batch to be complete.

File: tests/test_deprecated.py [hunk 2]
```diff
@@ -34,6 +34,7 @@ def test_tbd_remove_in_v0_8_0_trainer():
         'max_nb_epochs': 'max_epochs',
         'min_nb_epochs': 'min_epochs',
         'nb_sanity_val_steps': 'num_sanity_val_steps',
+        'fast_dev_run': 'unit_test',
```
Comment (awaelchli) [inline tests/test_deprecated.py]: This makes the test fail because in the trainer you set the max_epochs = 1 if self.unit_test, then the deprecated arg max_nb_epochs will have a different value (3) than max_epochs (1)

File: docs/source/debugging.rst [hunk 3]
```diff
@@ -13,7 +13,7 @@ argument of :class:`~pytorch_lightning.trainer.trainer.Trainer`)
 
 .. code-block:: python
 
-    trainer = pl.Trainer(fast_dev_run=True)
```
[the remark this answers is not in the document]
Comment (Gokkulnath) [reply docs/source/debugging.rst]: Sure will do the change

File: tests/test_deprecated.py (hunk 2 above)
Comment (Gokkulnath) [reply tests/test_deprecated.py]: Thanks a lot for helping. Shall i override the values in the test with the expected values i.e explicitly set values
`#### Override unit_test attributes for max_epochs and num_sanity_val_steps`
`kwargs['max_nb_epochs']=1 `
`kwargs['nb_sanity_val_steps']=1`

Comment (awaelchli) [reply tests/test_deprecated.py]: this should be ok since the test creates fixed values for the args anyway. 
@Borda is this ok?

Comment (awaelchli) [reply tests/test_deprecated.py]: > P,S : I am new to writing and modifying tests hence i am not sure how to proceed further.

I think what @williamFalcon meant was to make fast_dev_run also run on the test set. Then it would be a kind of unit test because it will check to run on train, val, and test data.

**Comment by Gokkulnath** [comment]
@Borda I think the PR is ready for review. Kindly let me know if any more changes are required.

File: pytorch_lightning/trainer/evaluation_loop.py [hunk 4]
```diff
@@ -160,7 +160,7 @@ class TrainerEvaluationLoopMixin(ABC):
     model: LightningModule
     num_test_batches: int
     num_val_batches: int
-    fast_dev_run: ...
+    unit_test: ...
```
Comment (Borda) [inline pytorch_lightning/trainer/evaluation_loop.py]:
```suggestion
    unit_test: bool
```

File: pytorch_lightning/trainer/trainer.py [hunk 5]
```diff
@@ -91,7 +91,8 @@ def __init__(
             overfit_pct: float = 0.0,
             track_grad_norm: int = -1,
             check_val_every_n_epoch: int = 1,
-            fast_dev_run: bool = False,
+            fast_dev_run=None,  # backward compatible, todo: remove in v0.8.0
+            unit_test: bool = False,
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py]: move this at the end as it kind of new and some users may use positional arguments

File: pytorch_lightning/trainer/trainer.py [hunk 6]
```diff
@@ -167,7 +168,12 @@ def __init__(
 
             check_val_every_n_epoch: Check val every n train epochs.
 
-            fast_dev_run: runs 1 batch of train, test  and val to find any bugs (ie: a sort of unit test).
+            fast_dev_run:
+                .. warning:: .. deprecated:: 0.7.0
+
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py]:
```suggestion
                .. warning:: .. deprecated:: 0.7.2
```

File: pytorch_lightning/trainer/trainer.py [hunk 7]
```diff
@@ -319,12 +325,18 @@ def __init__(
...
         self.shown_warnings = set()
 
-        self.fast_dev_run = fast_dev_run
-        if self.fast_dev_run:
+        self.unit_test = unit_test
+        # Backward compatibility, TODO: remove in v0.8.0
+        if fast_dev_run is not None:
+            warnings.warn("Argument `fast_dev_run` has renamed to `unit_test` since v0.5.0"
+                          " and this method will be removed in v0.8.0", DeprecationWarning)
+            self.fast_dev_run = fast_dev_run
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py]: all this move to deprecated API, also the warning gas wrong versions

File: pytorch_lightning/trainer/trainer.py [hunk 8]
```diff
@@ -319,12 +325,18 @@ def __init__(
...
+
+        if self.unit_test:
             self.num_sanity_val_steps = 1
             self.max_epochs = 1
             m = '''
-            Running in fast_dev_run mode: will run a full train,
+            Running in unit_test mode: will run a full train,
             val loop using a single batch
             '''
             log.info(m)
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py]: place the test directly, no need for `m = ...`

File: pytorch_lightning/trainer/training_loop.py [hunk 9]
```diff
@@ -180,7 +180,7 @@ class TrainerTrainLoopMixin(ABC):
     val_check_batch: ...
     num_val_batches: int
     disable_validation: bool
-    fast_dev_run: ...
+    unit_test: ...
```
Comment (Borda) [inline pytorch_lightning/trainer/training_loop.py]:
```suggestion
    unit_test: bool
```

File: tests/test_deprecated.py [hunk 10]
```diff
@@ -34,9 +34,14 @@ def test_tbd_remove_in_v0_8_0_trainer():
...
```
Comment (Borda) [inline tests/test_deprecated.py]: this is a deprecated test for v0.8 but you change is till v0.9 so move it to correct func

File: pytorch_lightning/trainer/training_loop.py (hunk 9 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/training_loop.py]: it's defined in trainer init, it should be inferred. I think keeping "..." is better.

File: docs/source/debugging.rst [hunk 11]
```diff
@@ -2,18 +2,18 @@ Debugging
 =========
 The following are flags that make debugging much easier.
 
-Fast dev run
-------------
+Unit test
+---------
 This flag runs a "unit test" by running 1 training batch and 1 validation batch.
```
Comment (awaelchli) [inline docs/source/debugging.rst]: does it also run 1 test batch now?

File: pytorch_lightning/trainer/training_loop.py (hunk 9 above)
Comment (Borda) [reply pytorch_lightning/trainer/training_loop.py]: but it is always `bool` type, right? then we shall be as specific as we can...

File: docs/source/debugging.rst (hunk 11 above)
Comment (Gokkulnath) [reply docs/source/debugging.rst]: yes. It will run for a single test batch.

File: tests/test_deprecated.py (hunk 10 above)
Comment (Gokkulnath) [reply tests/test_deprecated.py]: I have done the requested changes. Kindly review whenever possible.

File: CHANGELOG.md [hunk 12]
```diff
@@ -17,7 +17,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Changed
 
--
+- Renamed `fast_dev_run` to `unit_test` ([#1087](https://github.com/PyTorchLightning/pytorch-lightning/pull/1087))
 
 ### Deprecated
```
Comment (Borda) [inline CHANGELOG.md]: add note that `fast_dev_run` is deprecated since now

File: docs/source/debugging.rst [hunk 13]
```diff
@@ -2,18 +2,18 @@ Debugging
...
-This flag runs a "unit test" by running 1 training batch and 1 validation batch.
-The point is to detect any bugs in the training/validation loop without having to wait for
+Unit test
+---------
+This flag runs a "unit test" by running 1 training batch, 1 validation batch and 1 test batch.
+The point is to detect any bugs in the training/validation/test loop without having to wait for
 a full epoch to crash.
 
-(See: :paramref:`~pytorch_lightning.trainer.trainer.Trainer.fast_dev_run`
+(See: :paramref:`~pytorch_lightning.trainer.trainer.Trainer.unit_test`
```
Comment (Borda) [inline docs/source/debugging.rst]: this shall be `.. seealso::` @awaelchli right?

File: pytorch_lightning/trainer/deprecated_api.py [hunk 14]
```diff
@@ -87,3 +87,18 @@ def nb_sanity_val_steps(self, nb):
...
+        """Back compatibility, will be removed in v0.9.0"""
+        warnings.warn("Attribute `fast_dev_run` has renamed to `unit_test ` since v0.7.2"
+                      " and this method will be removed in v0.9.0", DeprecationWarning)
+        return self.unit_test
+
+    @fast_dev_run.setter
+    def fast_dev_run(self, unit_test):
+        """Back compatibility, will be removed in v0.9.0"""
+        warnings.warn("Attribute `fast_dev_run` has renamed to "
+                      "`unit_test` since v0.7.2"
```
Comment (Borda) [inline pytorch_lightning/trainer/deprecated_api.py]:
```suggestion
        warnings.warn("Attribute `fast_dev_run` has renamed to `unit_test` since v0.7.2"
```

File: pytorch_lightning/trainer/deprecated_api.py [hunk 15]
```diff
@@ -87,3 +87,18 @@ def nb_sanity_val_steps(self, nb):
                       "`num_sanity_val_steps` since v0.5.0"
                       " and this method will be removed in v0.8.0", DeprecationWarning)
         self.num_sanity_val_steps = nb
+
+    @property
+    def fast_dev_run(self):
```
Comment (Borda) [inline pytorch_lightning/trainer/deprecated_api.py]: you have mixed v0.8 and v0.9 deprecations as you can see above...
create a new deprecation mixin `TrainerDeprecatedAPITillVer0_9(ABC)`

File: tests/test_deprecated.py [hunk 16]
```diff
@@ -58,6 +58,23 @@ def test_tbd_remove_in_v0_9_0_module_imports():
...
+    kwargs = {k: (i + 1) for i, k in enumerate(mapping_old_new)}
+
+    trainer = Trainer(**kwargs)
+
+    for attr_old in mapping_old_new:
+        attr_new = mapping_old_new[attr_old]
+        assert kwargs[attr_old] == getattr(trainer, attr_old), \
+            'Missing deprecated attribute "%s"' % attr_old
+        assert kwargs[attr_old] == getattr(trainer, attr_new), \
+            'Wrongly passed deprecated argument "%s" to attribute "%s"' % (attr_old, attr_new)
```
Comment (Borda) [inline tests/test_deprecated.py]: as it become duplicated, move this to a particular function `_aasert_trainer_atribs(...)`

File: docs/source/debugging.rst (hunk 13 above)
Comment (Gokkulnath) [reply docs/source/debugging.rst]: I am not quite sure i understand the recommended change. I believe i need to add reference of Trainer module? Is my understanding correct ?

Comment (Borda) [reply docs/source/debugging.rst]: use `.. seealso::` intead `(See: `

File: tests/test_deprecated.py (hunk 16 above)
Comment (Gokkulnath) [reply tests/test_deprecated.py]: I have made the requested changes. Please review whenever possible

File: pytorch_lightning/trainer/trainer.py [hunk 17]
```diff
@@ -167,7 +169,12 @@ def __init__(
 
             check_val_every_n_epoch: Check val every n train epochs.
 
-            fast_dev_run: runs 1 batch of train, test  and val to find any bugs (ie: a sort of unit test).
+            fast_dev_run:
+                .. warning:: .. deprecated:: 0.7.2
+
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py]:
```suggestion
```

File: tests/test_deprecated.py [hunk 18]
```diff
@@ -58,6 +62,18 @@ def test_tbd_remove_in_v0_9_0_module_imports():
     from pytorch_lightning.logging.wandb import WandbLogger  # noqa: F402
 
 
+def test_tbd_remove_in_v0_9_0_trainer():
+    mapping_old_new = {
+        'fast_dev_run': 'unit_test',
+    }
+    # skip 0 since it may be interested as False
+    kwargs = {k: (i + 1) for i, k in enumerate(mapping_old_new)}
```
Comment (Borda) [inline tests/test_deprecated.py]: in this case, it is `bool` and the previous test tended to an arbitrary number...
consider changing/update to work with bools

Comment (Gokkulnath) [reply tests/test_deprecated.py]: you mean to set it to True/False instead of int right ?

Comment (Borda) [reply tests/test_deprecated.py]: you are right, as it is just one param, there is no need for this kind of heavy loop

File: pytorch_lightning/trainer/trainer.py (hunk 17 above)
Comment (Gokkulnath) [reply pytorch_lightning/trainer/trainer.py]: I see a blank line after each of the deprecated parameters. I thought it was a standard practice Should i remove the blank line after?

Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py]: @Borda The blank line is needed, otherwise sphinx will complain with a warning: "explicit markup ends without a blank line". Believe me or not it was a pain to find that out when I removed the warnings from the docs xD

Comment (Gokkulnath) [reply pytorch_lightning/trainer/trainer.py]: In that case, i will drop the recommendation. Feel free to suggest any more changes. If there is no other change, PR is ready to be merged

**Comment by Borda** [comment]
just opens an issue on Trains https://github.com/allegroai/trains/issues/119

**Comment by williamFalcon** [comment]
not sold at all. just wanted to have something that people “get” from looking at the name. but agree that unit_test is overloaded. @ethanperez

**Comment by Borda** [comment]
yeah, the Unit-test from sys eng. makes it a bit confusing... so what about "dry_run"?

**Comment by ethanjperez** [comment]
dry-run sounds good, but makes it sound to me like you won't save anything (like checkpoints), like with rsync's dry-run. Maybe --debug-run?

**Comment by Borda** [comment]
> dry-run sounds good, but makes it sound to me like you won't save anything (like checkpoints), like with rsync's dry-run. Maybe --debug-run?

so what about `simple-run`?

**Comment by jeremyjordan** [comment]
in my opinion, `fast_dev_run` is more informative - what's the motivation to change this?

**Comment by ethanwharris** [comment]
I'm with @jeremyjordan and @tullie on this. Unit_test to me implies testing some small unit of code rather than quickly leak testing the whole pipeline. Can't think of a good suggestion for a better name though ...

**Comment by Borda** [comment]
> I'm with @jeremyjordan and @tullie on this. Unit_test to me implies testing some small unit of code rather than quickly leak testing the whole pipeline. Can't think of a good suggestion for a better name though ...

do we have any third option as `unit_test` is soft.develop and `fast_dev_run` is not very descriptive

**Comment by Gokkulnath** [comment]
What about quick_test_pipelines?   Inspired from quickly leak testing the whole pipeline.

**Comment by tullie** [comment]
My favorite suggestion is probably debug-run but not totally convinced. What about “single_batch_run”?

**Comment by jeremyjordan** [comment]
there's a cost incurred when we change argument names, so i'd suggest we only do this when there's a clear improvement. what's not specific about `fast_dev_run`? in other words, what are we failing to communicate in that name? to me it's pretty clear, "oh we just want to do a quick run during development to make sure everything works"

**Comment by Borda** [comment]
> there's a cost incurred when we change argument names, so i'd suggest we only do this when there's a clear improvement. what's not specific about `fast_dev_run`? in other words, what are we failing to communicate in that name? to me it's pretty clear, "oh we just want to do a quick run during development to make sure everything works"

very much agree, let's hold this rename until we all agree on that the new name is better than the actual one... @Gokkulnath sorry for the confusion and very much appreciate your initiative, stay with us!

**Comment by Gokkulnath** [comment]
@Borda  No problem 👍

**Comment by Borda** [comment]
@PyTorchLightning/core-contributors are continuing in this renaming

**Comment by williamFalcon** [comment]
@Gokkulnath thanks for the contribution! this is my fault for not having arrived at a better name early haha. Please continute to contribute :)
